### PR TITLE
[PTFE-583] ✨ Prevent creation of integration branches

### PR DESCRIPTION
**Pull request by gaspardmoindrot**

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,6 +1,11 @@
 # Change Log
 All notable changes to this project will be documented in this file.
 
+## [3.8.0] - 2023-06-13
+# Added
+- Introducing new option `create_integration_branches`
```
Comment (nootal) [inline CHANGELOG:6]: We also introduce a general setting about it. I would not focus on the name of the option, more in the change of behavior here.

File: bert_e/docs/USER_DOC.md [hunk 2]
```diff
@@ -170,6 +171,20 @@ target development branch. This code is then tested in the build pipeline,
...
+and by default they are automatically created when a pull request is
+opened.
+This behaviour can be changed with the `always_create_integration_branches`
+parameter in the bot settings:
+
+* *true* (default): integration branches are created automatically when a
+  pull request is opened.
+* *false*: users will be required to explicitly request the creation
+  of integration branches by adding a `/create_integration_branches`
+  comment in their pull request.
```
Comment (nootal) [inline bert_e/docs/USER_DOC.md:184]: or it will be set for them during other operations.

File: bert_e/docs/USER_DOC.md [hunk 3]
```diff
@@ -170,6 +171,20 @@ target development branch. This code is then tested in the build pipeline,
...
+  The above allow to temporize the creation of those branches as
```
Comment (nootal) [inline bert_e/docs/USER_DOC.md:185]:
```suggestion
  The above allows to temporize the creation of those branches as
```

File: bert_e/docs/USER_DOC.md [hunk 4]
```diff
@@ -170,6 +171,20 @@ target development branch. This code is then tested in the build pipeline,
...
+  in the review process requested changes as expected.
```
Comment (nootal) [inline bert_e/docs/USER_DOC.md:186]: Not sure that's relevant here. If we add more commits later to take care of requested changes, then the integration branches will just naturally evolve. Their roles and solve any conflict before moving to the queues, so it makes sense to always have them, unless it's too costly.

File: bert_e/templates/request_integration_branches.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{% extends "message.md" %}
+
+{% block title -%}
+Request integration branches
+{% endblock %}
+
+{% block message %}
+Waiting for integration branch creation to be requested by the user.
+
+To request integration branches, please comment on this pull request with the following command:
```
Comment (nootal) [inline bert_e/templates/request_integration_branches.md:10]: What about the fact we wanted other comments to be able to set this value to true?

**Review by nootal — DISMISSED** [review]

**Comment by tcarmet** [comment]
Converting the PR as draft as it was open too soon with no changes to the draft code initially open in #124

File: CHANGELOG [hunk 6]
```diff
@@ -1,6 +1,11 @@
 # Change Log
 All notable changes to this project will be documented in this file.
 
+## [3.8.0] - 2023-06-13
```
Comment (tcarmet) [inline CHANGELOG:4]: I created a new development/3.9 branch and redirected your pr to it, as 3.8 has already been released and we're adding some feature into bert-e

```suggestion
## [3.9.0] - 2023-07-20
```

File: bert_e/workflow/gitwaterflow/integration.py [hunk 7]
```diff
@@ -136,6 +136,24 @@ def update(wbranch, source):
...
 
 
+def check_integration_branches(job):
+    """Check if the integration branches can be created."""
+
+    approvals = set(job.pull_request.get_approvals())
+    approved_by_author |= job.pull_request.author in approvals
+    if (approved_by_author is True and
+        len(job.git.cascade.dst_branches) > 1):
+        create_integration_branches()
```
Comment (tcarmet) [inline bert_e/workflow/gitwaterflow/integration.py:146]: integration branches will be created later on, this function shouldn't be called here.

File: README.md [hunk 8]
```diff
@@ -30,7 +30,7 @@ $ tox -e run
 A series of test scenario will be executed locally using mocks.
 
 ```shell
-$ tox -e tests
+$ tox run -e tests
```
Comment (tcarmet) [inline README.md:33]: I think I made some unnecessary changes here
```suggestion
$ tox -e tests
```

File: bert_e/docs/USER_DOC.md (hunk 4 above)
Comment (gaspardmoindrot) [reply bert_e/docs/USER_DOC.md:186]: I disagree with this point; I think it makes sense to keep this sentence. What do you think @tcarmet ?

File: bert_e/templates/request_integration_branches.md (hunk 5 above)
Comment (gaspardmoindrot) [reply bert_e/templates/request_integration_branches.md:10]: This message is displayed for people who need to type this command. In the case where someone writes `approve` at the beginning of their PR, they will instead receive a predefined message template that says: 'Integration branches have been created, congratulations! Here they are: ... blablabla.' They will never see this specific template, so I don't think it's necessary to mention it.

File: bert_e/tests/test_bert_e.py [hunk 9]
```diff
@@ -1159,6 +1159,90 @@ def test_comments_without_integration_pull_requests(self):
...
+build_key: pre-merge
+required_leader_approvals: 0
+required_peer_approvals: 1
+always_create_integration_branches: false
+admins:
+  - {admin}
+""" # noqa
+        options = self.bypass_all_but(['bypass_build_status'])
+        pr = self.create_pr('feature/TEST-0069', 'development/10.0')
+        self.handle(pr.id, settings=settings, options=options)
```
[the remark this answers is not in the document]
Comment (gaspardmoindrot) [reply bert_e/tests/test_bert_e.py:1405]: I can't figure out how to do it. Do you happen to have any ideas or suggestions? Thanks in advance

[the remark this answers is not in the document]
Comment (tcarmet) [reply bert_e/tests/test_bert_e.py:1405]: You did it just below, `with self.assertRaises(<add expected exceptions>)`

File: bert_e/docs/USER_DOC.md (hunk 4 above)
Comment (tcarmet) [reply bert_e/docs/USER_DOC.md:187]: Depends, why do you think it make sense to keep this sentence?

File: bert_e/tests/test_bert_e.py [hunk 10]
```diff
@@ -1159,6 +1159,264 @@ def test_comments_without_integration_pull_requests(self):
...
+
+        options = self.bypass_all
+        with self.assertRaises(exns.SuccessMessage):
+            self.handle(
+                pr.id, settings=settings, options=options, backtrace=True)
+        
+        self.assertIn(
+            'I have successfully merged the changeset', self.get_last_pr_comment(pr))
+
+    def test_creation_integration_branch_by_approve_2(self):
```
Comment (tcarmet) [inline bert_e/tests/test_bert_e.py:1323]: Add this test case inside the above test as they are in the same theme. Just create another PR.

**Review by tcarmet — CHANGES_REQUESTED** [review]

File: bert_e/docs/USER_DOC.md (hunk 4 above)
Comment (gaspardmoindrot) [reply bert_e/docs/USER_DOC.md:187]: It just provides more details on why there is no need to specify `/create_integration_branches` when typing `approve` and `create_pull_requests`.

File: bert_e/workflow/gitwaterflow/integration.py [hunk 11]
```diff
@@ -146,6 +146,8 @@ def check_integration_branches(job):
 
     if (job.settings.always_create_integration_branches is False and
             job.settings.create_integration_branches is False and
+            job.settings.always_create_integration_pull_requests is False and
```
[the remark this answers is not in the document]
Comment (gaspardmoindrot) [reply bert_e/workflow/gitwaterflow/integration.py]: Done

File: bert_e/workflow/gitwaterflow/integration.py [hunk 12]
```diff
@@ -136,6 +136,31 @@ def update(wbranch, source):
...
+    approvals = set(job.pull_request.get_approvals())
+    if job.settings.approve:
+        approvals.add(job.pull_request.author)
+    approved_by_author = job.pull_request.author in approvals
+
+    always_create_integration = job.settings.always_create_integration_branches
+    create_integration = job.settings.create_integration_branches
+    always_create_integration_prs = job.settings.always_create_integration_pull_requests
+    create_prs = job.settings.create_pull_requests
+    multiple_dst_branches = len(job.git.cascade.dst_branches) <= 1
```
Comment (tcarmet) [inline bert_e/workflow/gitwaterflow/integration.py:151]: You can go deeper and merge the following case into one variable. 

- `always_create_integration` with `create_integration`
- `always_create_integration_prs` and `create_prs`

**Review by tcarmet — CHANGES_REQUESTED** [review]

File: bert_e/docs/USER_DOC.md (hunk 4 above)
Comment (gaspardmoindrot) [reply bert_e/docs/USER_DOC.md:187]: Yes, I had understood correctly. My comment remains the same. However, I will remove it as it may not be clear.

File: bert_e/tests/test_bert_e.py [hunk 13]
```diff
@@ -1159,6 +1159,229 @@ def test_comments_without_integration_pull_requests(self):
...
+build_key: pre-merge
+required_leader_approvals: 0
+required_peer_approvals: 0
+always_create_integration_branches: false
+admins:
+  - {admin}
+""" # noqa
+        options = self.bypass_all_but(['bypass_build_status', 'bypass_author_approval'])
+        pr = self.create_pr('feature/TEST-0069', 'development/4.3')
+
```
[the remark this answers is not in the document]
Comment (gaspardmoindrot) [reply bert_e/tests/test_bert_e.py:1290]: Done!

File: bert_e/workflow/gitwaterflow/integration.py (hunk 12 above)
Comment (gaspardmoindrot) [reply bert_e/workflow/gitwaterflow/integration.py:151]: Done

File: bert_e/templates/request_integration_branches.md [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+```
+/create_integration_branches
+```
+
+Alternatively, there's another way to accomplish this. Simply tag this PR with `/approve` or
+`/create_pull_requests`, and it will automatically create the integration branches
```
Comment (tcarmet) [inline bert_e/templates/request_integration_branches.md:17]:
```suggestion
Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.
```
